File: src/automatica.core/Automatica.WebNew/src/app/base/model/rule-template.ts
import { BaseModel, JsonFieldInfo, JsonProperty, Model, JsonPropertyName } from "./base-model"
import { RuleInterfaceTemplate } from "./rule-interface-template"

function sortBySortOrder(a: RuleInterfaceTemplate, b: RuleInterfaceTemplate) {
    return a.SortOrder - b.SortOrder;
};

@Model()
export class RuleTemplate extends BaseModel {

    @JsonProperty()
    ObjId: string;

    @JsonProperty()
    Name: string;

    @JsonProperty()
    Description: string;

    @JsonProperty()
    Key: string;

    @JsonProperty()
    Group: string;

    @JsonPropertyName("RuleInterfaceTemplate")
    Interfaces: RuleInterfaceTemplate[] = [];

    @JsonProperty()
    Width: number;

    @JsonProperty()
    Height: number;

    @JsonProperty()
    This2DefaultMobileVisuTemplate: string;

    protected getJsonProperty(): Map<string, JsonFieldInfo> {
        return void 0;
    }

    get Inputs(): RuleInterfaceTemplate[] {
        const interfaces = new Array<RuleInterfaceTemplate>();
        this.Interfaces.forEach(element => {
            if (element.InterfaceDirection.Key === "I") {
                interfaces.push(element);
            }
        });

        return interfaces;
    }

    get Outputs(): RuleInterfaceTemplate[] {
        const interfaces = new Array<RuleInterfaceTemplate>();
        this.Interfaces.forEach(element => {
            if (element.InterfaceDirection.Key === "O") {
                interfaces.push(element);
            }
        });

        return interfaces;
    }


    protected afterFromJson() {
        this.Interfaces.sort(sortBySortOrder);
    }

    public typeInfo(): string {
        return "RuleTemplate";
    }
}

File: src/automatica.core/Automatica.Core.EF/Models/RuleTemplate.cs
using System;
using System.Collections.Generic;
using Automatica.Core.Model;

namespace Automatica.Core.EF.Models
{
    public class RuleTemplate : TypedObject
    {
        public RuleTemplate()
        {
            RuleInterfaceTemplate = new List<RuleInterfaceTemplate>();
        }

        public Guid ObjId { get; set; }
        public DateTimeOffset CreatedAt { get; set; }
        public DateTimeOffset ModifiedAt { get; set; }
        public Guid? Owner { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public string Key { get; set; }
        public string Group { get; set; }
        public float Width { get; set; }
        public float Height { get; set; }

        public Guid This2DefaultMobileVisuTemplate { get; set; }

        public List<RuleInterfaceTemplate> RuleInterfaceTemplate { get; set; }


        public VisuObjectTemplate This2DefaultMobileVisuTemplateNavigation { get; set; }
    }
}
